### ISSUE TITLE
Uploading videos

### ISSUE THREAD

**Issue by deanpcmad** [sferik/x-ruby#8]
With the current media upload example, when trying to upload a video, the API returns a `413 Payload Too Large` error.

Looking at the Upload API, for larger files they require [chunked media uploads](https://developer.twitter.com/en/docs/twitter-api/v1/media/upload-media/uploading-media/chunked-media-upload).

This requires multiple API requests to:

- Initialize the upload using the INIT command
- Upload each chunk of bytes using the APPEND command
- Complete the upload using the FINALIZE command

I've tried sending the required init params, but I keep getting `401 Unauthorized` errors. So I'm not sure if it's something I'm doing wrong, it's an issue on Twitter's end or if something needs tweaking in this library

```ruby
upload_client = X::Client.new(base_url: "https://upload.twitter.com/1.1/", **x_credentials)
upload_client.post("media/upload.json?command=INIT&media_type=video/mp4&total_bytes=4430752")   
```

Any ideas? Or could you add an additional example? Thanks

**Comment by sferik** [comment]
@deanpcmad Thank you for opening this issue! As I started looking into this and found a couple bugs in the library.

The main bug is that I’m not escaping query parameters from URLs. I’ve already [fixed](https://github.com/sferik/x-ruby/commit/56d6bd28f8132a9db42e661f5855b4c550c8878a) this bug and will release gem version 0.9.1 shortly. In the mean time, you can apply this patch to your code as a workaround (as soon as 0.9.1 is released, you can revert back to your original code):

```diff
 upload_client = X::Client.new(base_url: "https://upload.twitter.com/1.1/", **x_credentials)
-upload_client.post("media/upload.json?command=INIT&media_type=video/mp4&total_bytes=443075")
+upload_client.post("media/upload.json?command=INIT&media_type=video%2Fmp4&total_bytes=443075")
```

**Comment by deanpcmad** [comment]
Ah neat, thanks!

**Comment by sferik** [comment]
@imack I’ve added a new `X::MediaUpload` helper and updated the [examples](https://github.com/sferik/x-ruby/blob/main/examples/chunked_media_upload.rb) to show how to use it. This should allow you to greatly simplify your code. This will be included in the next release of this gem (version 0.10.0).

Note: this helper will automatically split videos into 5 MB chunks. You can adjust the chunk size, if you like.

**Comment by deanpcmad** [comment]
So I've just had chance to test this out, but I'm having encoding issues when trying to upload through this.

I've tried both downloading the video as a TempFile in my app and by downloading via Firefox and setting the path, but it always returns this error:

```
gems/x-0.12.0/lib/x/media_uploader.rb:75:in `write': "\xDA" from ASCII-8BIT to UTF-8 (Encoding::UndefinedConversionError)
```

Any ideas?

**Comment by sferik** [comment]
@deanpcmad This should be fixed by https://github.com/sferik/x-ruby/pull/13. I’ve just released version 0.12.1, which includes this fix.
